llvm/lib/Transforms/Yk/BasicBlockTracer.cpp

```diff
@@ -0,0 +1,70 @@
+#include "llvm/Transforms/Yk/BasicBlockTracer.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Module.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+
+#define DEBUG_TYPE "yk-basicblock-tracer-pass"
+
+using namespace llvm;
+
+namespace llvm {
+void initializeYkBasicBlockTracerPass(PassRegistry &);
+} // namespace llvm
+
+namespace {
+struct YkBasicBlockTracer : public ModulePass {
+  static char ID;
+  Function *externalFunc = NULL;
+
+  YkBasicBlockTracer() : ModulePass(ID) {
+    initializeYkBasicBlockTracerPass(*PassRegistry::getPassRegistry());
+  }
+
+  bool doInitialization(Module &M) override {
+    LLVMContext &Context = M.getContext();
+    if (externalFunc == NULL) {
+      Type *returnType = Type::getVoidTy(Context);
+      Type *functionIndexArgType = Type::getInt32Ty(Context);
+      Type *blockIndexArgType = Type::getInt32Ty(Context);
+
+      FunctionType *FType = FunctionType::get(
+          returnType, {functionIndexArgType, blockIndexArgType}, false);
+      externalFunc = Function::Create(FType, GlobalVariable::ExternalLinkage,
+                                      YK_TRACE_FUNCTION, M);
+      return true;
+    }
+    return false;
+  }
+
+  bool runOnModule(Module &M) override {
+    int functionIndex = 0;
+    for (Module::iterator F = M.begin(), E = M.end(); F != E; ++F) {
+      IRBuilder<> builder(M.getContext());
+      int blockIndex = 0;
+      for (Function::iterator BB = F->begin(), E = F->end(); BB != E; ++BB) {
```

Comment:
I don't feel strongly about this, but `++x` is much rarer in code than `x++`, so I'm always suspicious when I see `++x`. I don't think any of the instances in this function are changed whether they're `++x` or `x++` so maybe we should use the more common of the two?

That said, if LLVM tends to prefer `++x` then ignore what I've said!

Reply:
Agree that `++x` is rarer than `x++`.

I usually use the prefix `++x` operator because I don't like that with postfix `x++` the initial value of `x` is used as the return value rather than the incremented one. 

When I look at LLVM code base, it seems to use a prefix `++` operator, random example: [Mips16HardFloat.cpp#L513](https://github.com/ykjit/ykllvm/blob/main/llvm/lib/Target/Mips/Mips16HardFloat.cpp#L513)

A range-based loop would make more sense here since we do not use the indices.
👉 e03d8f56b5bbd988a1f7a01cbef6c2ac9797515f

Reply:
Even better!